Source/DafnyLanguageServer.Test/Various/ConcurrentInteractionsTest.cs

```diff
@@ -73,6 +84,86 @@ public class ConcurrentInteractionsTest : DafnyLanguageServerTestBase {
       Assert.AreEqual(1, diagnostics.Length);
     }
 
+    [TestMethod, Timeout(MaxTestExecutionTimeMs)]
+    public async Task VerificationErrorDetectedAfterCanceledSave() {
+      // Create a document that'll be slightly slow to verify
+      var source = @"
+method Multiply(x: bv10, y: bv10) returns (product: bv10)
+  requires y >= 0 && x >= 0
+  decreases y
+  ensures product == x * y && product >= 0
+{
+  if y == 0 {
+    product := 0;
+  } else {
+    var step := Multiply(x, y - 1);
+    product := x + step;
+  }
+}".TrimStart();
+      var failSource = @"method Contradiction() { assert false; }";
+      await SetUp(new Dictionary<string, string>() {
+        { $"{DocumentOptions.Section}:{nameof(DocumentOptions.Verify)}", nameof(AutoVerification.OnSave) }
+      });
+      var documentItem = CreateTestDocument(source);
+      await client.OpenDocumentAndWaitAsync(documentItem, CancellationTokenWithHighTimeout);
+
+      DidChangeTextDocumentParams MakeChange(int? version, Range range, string text) {
+        return new DidChangeTextDocumentParams {
+          TextDocument = new OptionalVersionedTextDocumentIdentifier {
+            Uri = documentItem.Uri,
+            Version = version
+          },
+          ContentChanges = new[] {
+            new TextDocumentContentChangeEvent {
+              Range = range,
+              Text = text
+            }
+          }
+        };
+      }
+
+      // Add a space in the document
+      var change1 = MakeChange(documentItem.Version + 1, new Range((0, 6), (0, 6)), " ");
+      client.DidChangeTextDocument(change1);
+      // Save and don't wait, so the next save will interrupt and cancel verification
+      Thread.Sleep(1);
+      client.SaveDocument(documentItem);
+
+      // Remove the space, and use a non-consecutive version to test that the server doesn't drop the change
+      var change2 = MakeChange(documentItem.Version + 4, new Range((0, 6), (0, 7)), "");
+      client.DidChangeTextDocument(change2);
+      // Save and don't wait, so the next save will interrupt and cancel verification
+      Thread.Sleep(1);
```

Comment:
Could you replace calls to `Thread.Sleep` with `await Task.Delay` ?

Also, can we avoid using sleep by waiting on something else? What do you want to happen during the sleep? Tests that depend on sleeps are often unstable.

Reply:
The issue here is that cancelling the operation too quickly leads to an exception within the OmniSharp JSON parsing code. It isn't prepared for an operation to be cancelled before it has even finished parsing it. Fixing that would be ideal, but would probably require a change to the OmniSharp libraries. I'll make a note of it in a comment, though.

And, yes, I think `await Task.Delay` seems better. I'm still learning the C# concurrency API. :)

Reply:
> The issue here is that cancelling the operation too quickly leads to an exception within the OmniSharp JSON parsing code

Is that a bug in our code then, for cancelling code which shouldn't be cancelled? Maybe we can use different cancellationTokens to allow only cancelling the part that works well?

Also, do you happen to have more information on this OmniSharp JSON parsing exception? I might have been running into it without understanding it. I've seen this in some build runs:

```csharp
  Standard Output Messages:
 crit: OmniSharp.Extensions.JsonRpc.InputHandler[0]
       unhandled exception
       System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'start')
          at System.Buffers.BuffersExtensions.ToArray[T](ReadOnlySequence`1& sequence)
          at OmniSharp.Extensions.JsonRpc.InputHandler.ProcessInputStream(CancellationToken cancellationToken)
          at OmniSharp.Extensions.JsonRpc.InputHandler.ProcessInputStream(CancellationToken cancellationToken)
          at OmniSharp.Extensions.JsonRpc.InputHandler.<Start>b__26_0()
```

Reply:
The exception I had seen was in related code but was a different exception with a different message. In an attempt to get the exact message, I tried to remove the delays and reproduce the issue, and after 10 runs it didn't show up again. There have been a couple of other changes to the test, and changes in the rest of the code base merged in since then, including some in the test infrastructure.

Regardless, since I haven't been able to reproduce it, I'm going to remove the delays altogether. I'm much happier without them. If the issue shows up in the future, we can dig into why it's happening.